application/comit_node/src/swap_protocols/rfc003/state_store.rs

```diff
@@ -28,38 +27,37 @@ pub struct InMemoryStateStore<K: Hash + Eq> {
 }
 
 impl<K: Hash + Eq + Clone + Send + Sync + 'static> StateStore<K> for InMemoryStateStore<K> {
-    fn insert<R: Role>(
-        &self,
-        key: K,
-        state: SwapStates<R>,
-    ) -> Result<Arc<dyn SaveState<R>>, Error> {
+    fn new_save_state<R: Role>(&self, key: K) -> Result<Arc<dyn SaveState<R>>, Error> {
         let mut states = self.states.lock().unwrap();
 
         if states.contains_key(&key) {
             return Err(Error::DuplicateKey);
         }
 
-        let state = Arc::new(RwLock::new(state));
-
+        let state = Arc::new(RwLock::new(None));
         let value: Box<dyn Any + Send + Sync> = Box::new(state.clone());
-        let _old = states.insert(key, value);
+        let _ = states.insert(key, value);
 
         Ok(state)
     }
 
     fn get<R: Role>(&self, key: &K) -> Result<Option<SwapStates<R>>, Error> {
         let states = self.states.lock().unwrap();
-        Ok(states.get(key).map(|state| {
-            let state = state.downcast_ref::<Arc<RwLock<SwapStates<R>>>>().unwrap();
+        Ok(states.get(key).and_then(|state| {
+            let state = state
+                .downcast_ref::<Arc<RwLock<Option<SwapStates<R>>>>>()
+                .unwrap();
```

Comment:
🙃 we are returning a result, should we really `unwrap`?

Reply:
True. It was like this before but it's wrong. IMO We should return `None` in this case rather than an error though.